### ISSUE TITLE
Loosing tab-groups when opening / closing windows

### ISSUE THREAD

**Issue by skela** [Quicksaver/Tab-Groups#174]
I originally wrote about my issues here in this review: https://addons.mozilla.org/en-US/firefox/addon/tab-groups-panorama/reviews/788903/

I should have added them here sorry, but had a feeling that this might have been "as designed", which is why I went with the review in the first place.

So I've got 1 window open with a bunch of tab groups, all neatly named, and organised.
If I quit firefox, and go back in, everything reloads fine, no problem.

If I open a new window, and quit firefox, and go back in, everything reloads fine, no problem.

The issue arises if I open a new window, then accidentally close the window with all the groups. If I quit firefox, and go back in, all my tab groups are lost.

This is quite easy to do because its not always intuitive that 1 window has the tab-groups and other windows do not, or that each window can have a unique set of tab-groups.

You will know best what to do to avoid these kinds of issues, but here's 1 thought I had that might help; If you incorporate the concept of windows in the tab-groups selection UI, people could move 
tab-groups between windows if they so wanted. Also, you could de-couple tab-groups from window instances, so that in my situation, I would come back in, and I'd see my set of tab-groups not attached to any window, and I could then manually attach them, or close them at will. Let me know what you think, or if anything is unclear.

I personally put a lot of effort into the tab-groups, I name them and organise them and it helps me switch between projects easily. But the fact is that because of the windowing issues, I keep loosing my groups, and its frustrating.

**Comment by iShyboy** [comment]
You should use Session Manager, and program a auto backup for your session, even in windows it could save your groups in a set time.
There's a lot of ppl talking about sharing groups between windows, this could solve your issue, or you just want a prompt that shows up when more than one window its opened and have more than one group?

**Comment by skela** [comment]
I had a look at Session Manager, and its not really what I'm after here.
Yes, having all the tab-groups shared amongst windows, would probably fix the issue too.

**Comment by iShyboy** [comment]
Session manager would just prevent you from having a headache, since you
could just change for another backup, and that'll open all the windows from
that time, with the groups.

-- 
-
Cordialmente,
Juliandro de Sousa Vasconcelos

**Comment by Quicksaver** [comment]
Indeed, that is not quite what I understood from your review, as you have probably already guessed from my response there. ;) The gist of it is, windows and sessions are different things. Sessions can have multiple windows open (and closed), and you can only run one session at a time.

Sharing groups across windows is something I already want to do, although admittedly it's not very high on my priorities list, mostly because it is not something that is easily done, it's not just a matter of sharing groups across windows, it's also sharing tabs across windows. This... is complicated, tabs just aren't built that way. (I've changed the title of the issue to better reflect my plan here.)

I also agree that Session Manager may not be the most useful tool for your case. Although it's easy to get around the issue only with Firefox. When you close a window but still have another one open, you can get it back by clicking the History button (by default it's in the menu panel, the one that opens when you click the three bars / hamburger button at the top right). If there are any closed windows in your current session, they will appear in a "Recently closed windows" section in there.

If you're still running the same session as when you lost your groups (when you first wrote the review), you may even still be able to get back the groups you lost then.

**Comment by iShyboy** [comment]
Welp, I guess session manager is just too dependable.
![SM](http://i.imgur.com/pPXIoxP.png)
Even in this aspect, I dont remember if FF recover the closed windows in history, if he already closed FF and opened it again.

**Comment by Quicksaver** [comment]
> Welp, I guess session manager is just too dependable.

I'd be surprised if the leading session manager add-on didn't keep track of closed windows. ;)

> Even in this aspect, I dont remember if FF recover the closed windows in history, if he already closed FF and opened it again.

It does. At one point I had a few closed windows in my developer profile from so long before that I didn't even remember what their purpose was. They may still be cleaned after some time, although I don't have the specifics on that. But that's why I said it **may** be possible. :)

> I tried opening 2 FF window, the 2 with groups and tabs, and closing one, then quitting in another, when I started FF again, it OPENED THE 2 window perfectly fine, with tabs and groups. And I have just tab groups in that one profile.

You may just not have given it enough time to record the new session data to disk; this doesn't happen immediately on every change as to spare the disk from unnecessary heavy read-write cycles. Or it may be that Session Manager handles this aspect a little differently than native Firefox's session handler.

**Comment by iShyboy** [comment]
In the profile for the test, I just installed tab groups, and FF **should** save the session to disk in a proper quit. Anyway the issue it's just a bit irrelevant, if FF does it in a native way.

**Comment by skela** [comment]
Thanks, the Show Recently Closed Windows is exactly the kind of workaround I needed. As far as I can tell, there's no issue then, and this is more of a feature request that I think would make this plugin even more awesome than it already is :) Keep up the good work guys, the fact that Firefox removed this behavior is beyond me, but I reckon this is a good opportunity to improve some of the reasons why I think it (the functionality) failed to gain popularity in the first place.

**Comment by Quicksaver** [comment]
@skela since this isn't actually a bug and everything is working as expected, especially since you can get back the closed windows, could I ask you to update your review and rating? :)

**Comment by skela** [comment]
@Quicksaver sure I've updated it, gave it 4 stars , will give it 5 when Shared groups across windows is completed 👍

**Comment by Quicksaver** [comment]
Not to sound discouraging, but that's a somewhat unrealistic expectation. It may not be possible to do something like that at all. As I said, tabs themselves just aren't built that way. I will definitely work on this (sometime) but if it turns out it requires some heavy rewrites on Firefox's own mechanisms that handle tabs, I probably won't implement it as it would mean far too many changes for honestly not that big of a gain.

**Comment by WorldsEndless** [comment]
Just a +1 on this. I don't really understand Firefox' handling of new windows anyway, but it would be great to share groups across them (as if a new window was just an independently displayed tab)

**Comment by wrossmann** [comment]
Another possible aspect/feature/preliminary fix to this would be facilitating moving tab groups between windows.

Really, all I want right now is to move the group from my second window into the first and go back to one window. I'm not by any means familiar with FF internals, but maybe it's possible to make use of the existing mechanism that FF uses to move tabs between windows to get around the fact that tabs cannot be "shared" by windows.

**Comment by spinda** [comment]
I'll throw another "yes please" in here. This would be really helpful for multi-monitor setups.
